JENKINS-53284 Fix warning message for changing workspace directories

charanbir opened the pull request:
See [JENKINS-53284](https://issues.jenkins-ci.org/browse/JENKINS-53284).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* [JENKINS-53284](https://issues.jenkins-ci.org/browse/JENKINS-53284): only warn about changing workspace or builds directories on non fresh startups

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers


<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

charanbir committed:
JENKINS-53284 Fix warning message for changing workspace directories

charanbir commented:
This is my first open source contribution commit ever. Please let me know if I missed anything. Thank you.

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3033,9 +3033,16 @@ private void setBuildsAndWorkspacesDir() throws IOException, InvalidBuildsDir {
-        if (newWorkspacesDir != null && !workspaceDir.equals(newWorkspacesDir)) {
-            LOGGER.log(Level.WARNING, "Changing workspaces directories from {0} to {1}. Beware that no automated data migration will occur.",
-                       new String[]{workspaceDir, newWorkspacesDir});
+        if (newWorkspacesDir != null) {
```
NB: can be written more concisely as
```
Level level = workspaceDir.equals(newWorkspacesDir) ? Level.WARNING : Level.INFO;
```

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3033,9 +3033,16 @@ private void setBuildsAndWorkspacesDir() throws IOException, InvalidBuildsDir {
+        	else {
+        		level = Level.WARNING;
+        	}
+            LOGGER.log(level, "Changing workspaces directories from {0} to {1}. Beware that no automated data migration will occur.",
```
there is something wrong with the indentation here.

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3033,9 +3033,16 @@ private void setBuildsAndWorkspacesDir() throws IOException, InvalidBuildsDir {
+        		level = Level.INFO;
+        	}
+        	else {
+        		level = Level.WARNING;
```
won't this cause the WARNING message to always be logged on all startups except the first one?

jtnord commented:
Hi @charanbir, 
The associated unit test (jenkins.model.JenkinsBuildsAndWorkspacesDirectoriesTest) is failing, possibly due to https://github.com/jenkinsci/jenkins/pull/3728#discussion_r232275121?

charanbir commented:
Thank you for your input and guidance. I will make the necessary changes.

charanbir committed:
JENKINS-53284 Fix warning message for changing workspace directories

charanbir commented:
@jtnord Changes made

jtnord commented:
@charanbir Can I suggest you write a unit test for the expected behaviour, there is something not quite right with your code which will become obvious if you have a unit test.

batmat commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3034,7 +3034,8 @@ private void setBuildsAndWorkspacesDir() throws IOException, InvalidBuildsDir {
         String newWorkspacesDir = SystemProperties.getString(WORKSPACES_DIR_PROP);
         if (newWorkspacesDir != null && !workspaceDir.equals(newWorkspacesDir)) {
-            LOGGER.log(Level.WARNING, "Changing workspaces directories from {0} to {1}. Beware that no automated data migration will occur.",
+        	Level level = workspaceDir.equals(DEFAULT_WORKSPACES_DIR) ? Level.INFO : Level.WARNING;
```
I think this will not work. If the user changes it once, then again, the warning will not be shown when it should be.

I agree with James that writing a test for this is important here. This will make it obvious that it works or not, and will make much easier for you @charanbir to test and iterate on making sure your change does work.

jtnord replied:
when will this be set to `INFO` and when will it be set to `WARNING`, what is the test condition in the surrounding block?

batmat reviewed (changes requested):
Awesome work! Thanks a lot for chiming in @charanbir !

We're almost there :smile_cat:. Please see to add an automated test, likely a new case in the existing `JenkinsBuildsAndWorkspacesDirectoriesTest`.
You will see a `LoggerRule` is already in use there with example on how to check if a log was issued or not.

Thanks again!
![](https://media.giphy.com/media/GZiMCzG87E2QM/giphy.gif)

charanbir committed:
JENKINS-53284 Fix warning message for changing workspace directories

charanbir committed:
Merge remote-tracking branch 'upstream/master' into JENKINS-53284

batmat commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -258,6 +292,15 @@ private boolean logWasFound(String searched) {
                 .anyMatch(record -> record.getMessage().contains(searched));
     }
 
+	private boolean logLevelWasFound(String searched, Level level) {
```
`logWasFoundAtLevel` would probably better reflect what it does.
Also, I think `logWasFound` should delegate to it.

batmat commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -258,6 +292,15 @@ private boolean logWasFound(String searched) {
+				.getRecords()
+				.stream()
+				.filter(record -> record.getMessage().contains(searched)
+						&& record.getLevel().equals(level))
```
For lambda clarity/declarativeness, I'd recommend putting two filters instead of using `&&`:

```java
.filter(record -> record.getMessage().contains(searched))
.filter(record -> record.getLevel().equals(level))
```

batmat reviewed (commented):
I built that PR locally and tested it. For me it does not work as expected or I'm missing something.
The following use case issues the log as `INFO` when it should be a WARNING:

```
# move under a fresh empty directory
 # no sysprop specified
JENKINS_HOME=$PWD java -jar /path/to/jenkins.war --httpPort=8080
...
 # Following startup logs as `INFO` when it should be a WARNING
JENKINS_HOME=$PWD java -Djenkins.model.Jenkins.workspacesDir=bluh -jar /path/to/jenkins.war --httpPort=8080
...
 # Following startup logs as `WARNING` correctly, since bluh was changed to blah \o/
JENKINS_HOME=$PWD java -Djenkins.model.Jenkins.workspacesDir=blah -jar /path/to/jenkins.war --httpPort=8080
```

Also, seems like the `buildsDir` sysprop is not handled in this PR, but definitely fine by me to do it in a dedicated PR after this one gets merged.

Thanks a lot for your efforts @charanbir !

charanbir committed:
Merge branch 'master' of https://github.com/charanbir/jenkins into JENKINS-53284

charanbir committed:
JENKINS-53284 Fix warning message for changing workspace directories

charanbir commented:
Alright, I'll try to fix it. Please bear with me.

charanbir committed:
Merge branch 'master' of https://github.com/charanbir/jenkins into JENKINS-53284

charanbir committed:
JENKINS-53284 Fix warning message for changing workspace directories

charanbir commented:
@batmat 

- Thank you for reviewing. I made the changes you suggested, including the verbiage. 
- In all your commands you have initialized JENKINS_HOME yourself, so that value will never be null.
- I couldn't use the 2 variables `newWorkspacesDir` and `workspaceDir` to determine if it was a fresh startup so I created a boolean variable. This works and I tested all cases mentioned in the above comments. I also used the same logic for `buildsDir`.
- Please let me know if there are any concerns or changes.

batmat commented:
@charanbir sorry for the lag, I've been busy on a few other things. I didn't forget this and hope to get to it again very soon. Thanks a lot for the work!

oleg-nenashev commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -66,6 +69,31 @@ private void clearSystemProperties() {
                 .forEach(System::clearProperty);
     }
+    
+	@Issue("JENKINS-53284")
```
tabs vs. spaces

oleg-nenashev reviewed (changes requested):
`freshStartup` will not work correctly if you keep it in `config.xml` like it is done in this pull request:

* It is NOT guaranteed that fresh-start instances have no job data added in JENKINS_HOME
* It will suppress the message on the first upgrade to the version

Since the warning is still present as INFO, I request changes only for the data storage. It should be rather a separate marker file IMO

batmat commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -258,6 +286,13 @@ private boolean logWasFound(String searched) {
                 .anyMatch(record -> record.getMessage().contains(searched));
     }
 
+	private boolean logWasFoundAtLevel(String searched, Level level) {
```
NIT: `logWasFound` body above should probably be rewritten to `logWasFoundAtLevel(searched, Level.ALL)`

batmat replied:
my bad, ^ is wrong, because it's an exact level match. So it's not fully as straightforward as I was thinking.

batmat commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -258,6 +286,13 @@ private boolean logWasFound(String searched) {
+		return loggerRule.getRecords().stream()
+				.filter(record -> record.getMessage().contains(searched))
+				.filter(record -> record.getLevel().equals(level))
+				.collect(Collectors.toList()).size() > 0;
```
NIT: see `anyMatch()` above. It's somehow unimportant here in test context, but it **is** important in general. Because what the code here is doing necessitates to go computing the whole list size. 

I.e. if `Collectors.toList()` is actually going to compute a list of billions of elements, it could take a lot of time, when anyMatch will just stop on the first match and yield the return result.

batmat commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -66,6 +69,31 @@ private void clearSystemProperties() {
+		});
+
+	}
+    
```
Your code formatter seems to have an issue :-)
![image](https://user-images.githubusercontent.com/223853/51172485-9df80300-18b3-11e9-9e30-4ed248a2660a.png)

batmat reviewed (approved):
LGTM from local testing.

batmat committed:
Reformat new tests (tabs vs. space)

I, for one, prefer tabs, but spaces is the practice in place in Jenkins,
hence aligning.

batmat committed:
Replace attribute by marker file for Jenkins startup

Goal is to be especially more compatible with tools like JCasC that will
rewrite the config.xml and hence likely remove the freshStartup attribute.

So we store this information as a separate marker file at the end of the
Jenkins startup process to avoid losing it.

batmat commented:
@charanbir I've added the commits requested by Oleg myself, because I feel like I owe you some time given how long you've waited for this to be looked at again :-|. I hope you do not mind.

@oleg-nenashev I believe I addressed your concern, and used your recommendation to use a marker file instead. Care to have another look?

I've also retested locally, and this works as expected (first startup, `INFO` only, `WARN` next times if one changes the buildsDir or workspacesDir value -- note: this is/was anyway also JUnit tested).

MRamonLeon commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -838,7 +849,7 @@ protected Jenkins(File root, ServletContext context, PluginManager pluginManager
         JenkinsJVMAccess._setJenkinsJVM(true); // set it for unit tests as they will not have gone through WebAppMain
         long start = System.currentTimeMillis();
-
+        STARTUP_MARKER_FILE = new FileBoolean(new File(root, ".lastStarted"));
```
Is this file ever removed? In a classic Jenkins instance it will always stay there after the first start. Doesn't it involves too much (adding an extra file for ever) for just a level in the warning? Isn't there any other way to catch we are in a fresh startup? I.e: the former build or workspace directories already exists or something like that.

batmat replied:
I believe 1) adding marker file is a simple enough solution and commonly used in Jenkins codebase (see the `secret.key.not-so-secret` file for instance and 2) this _could_ be useful for other code. Low-tech is sometimes also good too IMO: if something goes wrong, we can ask user to delete (or create) this file to trigger/revert to some other behavior.

MRamonLeon replied:
I agree low-tech is good, almost always. I was just asking if there were any other approach, with the existing infra (files, directories, status files)

MRamonLeon commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -67,6 +70,29 @@ private void clearSystemProperties() {
+        story.then(step -> {
+            assertFalse(logWasFound(LOG_WHEN_CHANGING_WORKSPACES_DIR));
+            setWorkspacesDirProperty("testdir1");
+        });
```
Shouldn't you test the log is found at level INFO?

batmat replied:
At that stage, we have already started (i.e. **without** any flag), and about to restart), so in that first case: **no** log at all is expected, so this way works fine.

As you can see below, on the second restart after having set the property: then yes we must check the level more specifically.

Does that clarify @MRamonLeon?

MRamonLeon replied:
Yes, I was asking to check when the message is raised in INFO level.

MRamonLeon reviewed (commented):
Some doubts

batmat commented:
@jtnord care to have a second look? Thanks!

MRamonLeon reviewed (approved):
Anyway, LGTM

batmat commented:
Retriggered a build, looked like an infra issue
```
Creating placeholder flownodes because failed loading originals.

GitHub has been notified of this commit’s build result

java.io.IOException: Tried to load head FlowNodes for execution Owner[Core/jenkins/PR-3728/36:Core/jenkins/PR-3728 #36] but FlowNode was not found in storage for head id:FlowNodeId 1:196
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.initializeStorage(CpsFlowExecution.java:678)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.onLoad(CpsFlowExecution.java:715)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getExecution(WorkflowRun.java:664)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.onLoad(WorkflowRun.java:526)
	at hudson.model.RunMap.retrieve(RunMap.java:225)
	at hudson.model.RunMap.retrieve(RunMap.java:57)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:501)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:483)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:381)
	at hudson.model.RunMap.getById(RunMap.java:205)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.run(WorkflowRun.java:901)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.get(WorkflowRun.java:912)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getFlowExecution(CpsStepContext.java:474)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getThreadGroupSynchronously(CpsStepContext.java:243)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.access$000(CpsStepContext.java:97)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext$1.call(CpsStepContext.java:264)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext$1.call(CpsStepContext.java:262)
	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

oleg-nenashev reviewed (approved):
Thanks to @batmat for addressing the change request

batmat commented:
I consider this ready for merge. 
I will likely merge it later today or tomorrow (if the PR build goes green, obviously), to possibly wait for more feedback (even if quite unlikely given this has been here for quite long already :-)).

Thanks!

jglick commented:
Rather than suppressing warnings, I would suggest just removing support for the old locations, which is long obsolete. As I noted in [JENKINS-21942](https://issues.jenkins-ci.org/browse/JENKINS-21942), contrary to expectations the old locations are used in _new_ installations under some circumstances.

daniel-beck commented:
@charanbir Please do not delete the pull request template.
@batmat Please restore the PR template in such cases, or don't merge nontrivial PRs without is.

daniel-beck commented:
Does this actually work as intended? The idea is to print a warning when the directory changes, but all this does is to make the first startup with any directory defined print a warning, and none of the subsequent startups, i.e. if the specified directory is changed between subsequent startups, no warning will be written.

batmat commented:
> Please restore the PR template in such cases, or don't merge nontrivial PRs without is.

@daniel-beck my bad. I generally do it, but in this case, seems like I missed, maybe because the PR had been running for quite long.

batmat commented:
> Does this actually work as intended?

AFAIR from my testing last week, yes. And there are automated tests also, so yes I think this works. 
I will anyway be able to confirm it soon: once the weekly is out, I plan to update Evergreen with this. I'll get back here if this actually does not work. 

I was even planning to use the Incremental release right away, but this is actually currently broken, so I had to wait for the weekly to deploy it :disappointed:. ([INFRA-1964](https://issues.jenkins-ci.org/browse/INFRA-1964))
